components/raftstore/src/store/worker/split_check.rs

```diff
@@ -284,28 +377,66 @@ where
             let mut size = 0;
             let mut keys = 0;
             let mut bucket_size: u64 = 0;
+            let mut bucket_range_idx = 0;
+            let mut skip_on_kv = false;
             while let Some(e) = iter.next() {
-                if host.on_kv(region, &e) {
+                if skip_on_kv && skip_check_bucket {
                     return;
                 }
+                if !skip_on_kv && host.on_kv(region, &e) {
+                    skip_on_kv = true;
+                }
                 size += e.entry_size() as u64;
                 keys += 1;
-                if host.enable_region_bucket() {
-                    bucket_size += e.entry_size() as u64;
-                    if bucket_size >= host.region_bucket_size() {
-                        bucket_keys.push(keys::origin_key(e.key()).to_vec());
-                        bucket_size = 0;
+                if !skip_check_bucket {
+                    let origin_key = keys::origin_key(e.key());
+                    if bucket_range_list.is_empty() {
+                        bucket_size += e.entry_size() as u64;
+                        if bucket_size >= host.region_bucket_size() {
+                            bucket.keys.push(origin_key.to_vec());
+                            bucket.size += bucket_size;
+                            bucket_size = 0;
+                        }
+                    } else {
+                        // find proper bucket range that covers e.key
+                        while bucket_range_idx < bucket_range_list.len()
+                            && origin_key >= bucket_range_list[bucket_range_idx].1.as_slice()
+                        {
+                            bucket_range_idx += 1;
+                            bucket_size = 0;
+                            if buckets.len() < bucket_range_idx {
+                                buckets.push(bucket);
+                                bucket = Bucket::default();
+                            }
+                        }
+                        if bucket_range_idx == bucket_range_list.len() {
+                            skip_check_bucket = true;
+                        } else if origin_key >= bucket_range_list[bucket_range_idx].0.as_slice() {
+                            // e.key() is between bucket_range_list[bucket_range_idx].0, bucket_range_list[bucket_range_idx].1
+                            bucket_size += e.entry_size() as u64;
+                            if bucket_size >= host.region_bucket_size() {
+                                bucket.keys.push(origin_key.to_vec());
+                                bucket.size += bucket_size;
+                                bucket_size = 0;
+                            }
+                        }
                     }
                 }
             }
 
+            if buckets.len() < bucket_range_list.len()
```

Comment:
Is there the invariant that `buckets.len() == bucket_range_list.len()`?

Reply:
I saw the assertion at https://github.com/tikv/tikv/pull/12132/files#diff-7a4718b9cddd27a4ef7e563024d5283cf14b4ba3abf26edc33b974b87680a5fbR4822.

Reply:
> Is there the invariant that `buckets.len() == bucket_range_list.len()`?

When bucket_range_list is None, it would still return one bucket object that contains all the split keys for the whole region.

Reply:
> I saw the assertion at https://github.com/tikv/tikv/pull/12132/files#diff-7a4718b9cddd27a4ef7e563024d5283cf14b4ba3abf26edc33b974b87680a5fbR4822. It would be better to assert it here.

fair.